File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-0.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        retVal = 0
        hold = set(s)

        for x in hold:
            count = 0
            front = 0
            for back in range(len(s)):
                if s[back] == x:
                    count += 1
                
                while (back - front + 1) - count > k:
                    if s[front] == x:
                        count -= 1
                    front += 1
                retVal = max(retVal, back-front+1)

        return retVal

File: Data Structures & Algorithms/time-based-key-value-store/submission-0.cpp
class TimeMap {
public:
    unordered_map<string, vector<pair<int, string>>> hold;

    TimeMap() {
        
    }
    
    void set(string key, string value, int timestamp) {
        hold[key].push_back({timestamp, value});
    }
    
    string get(string key, int timestamp) {
        string retVal = "";
        auto& values = hold[key];
        int front = 0;
        int back = values.size() - 1;

        while (front <= back) {
            int mid = front + (back - front) / 2;

            if (values[mid].first <= timestamp) {
                retVal = values[mid].second;
                front = mid + 1;
            } else {
                back = mid - 1;
            }
        }
        return retVal;
    }
};
